# Chapter 3: Data exploration

In [31]:
import numpy as np
import pandas as pd

## 3.1 Introduction & Problem Setting

As we discussed in the first lesson, the most important goal of data processing is to ensure that the data we work with is clean and reliable. To achieve this, we need to explore our dataset thoroughly, understand its structure, and identify any issues or "cracks" in the data. Only then can we create a high-quality dataset ready for analysis. Think of it like being a detective at a crime scene – you examine every clue carefully to solve the mystery! 🔍

## 3.2 Explore the content of the dataset

Let's dive in! Today, we will work with the Titanic dataset 🚢, a well-known dataset that contains information about the passengers on the Titanic when it sank. We load the dataset and take a quick look at the data.

In [32]:
titanic = pd.read_csv("titanic_uncleaned.csv", sep = ";")
titanic.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,No,3.0,"Braund, Mr. Owen Harris",male,NaN,1.0,0.0,A/5 21171,7.25,NaN,S
1,2,Yes,1.0,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1.0,0.0,PC 17599,712.833,C85,C
2,3,Yes,3.0,"Heikkinen, Miss. Laina",female,26.0,0.0,0.0,STON/O2. 3101282,7.925,NaN,S
3,4,Yes,1.0,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1.0,0.0,113803,53.1,C123,S
4,5,No,3.0,"Allen, Mr. William Henry",male,35.0,0.0,0.0,373450,8.05,NaN,S


Here, we load the Titanic dataset from a CSV file. The `sep=';'` parameter specifies that the values are separated by semicolons instead of commas, which is common in some European data formats.

#### Data Dictionary

| Variable | Definition | Key |
|----------|------------|-----|
| survival | Survival | 0 = No, 1 = Yes |
| pclass | Ticket class | 1 = 1st, 2 = 2nd, 3 = 3rd |
| sex | Sex |  |
| Age | Age in years |  |
| sibsp | # of siblings / spouses aboard the Titanic |  |
| parch | # of parents / children aboard the Titanic |  |
| ticket | Ticket number |  |
| fare | Passenger fare |  |
| cabin | Cabin number |  |
| embarked | Port of Embarkation | C = Cherbourg, Q = Queenstown, S = Southampton |

Another way to look at some data is by using .sample() instead of .head(). You can pass a number to determine how many random samples to display.

In [33]:
titanic.sample(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
783,784,No,3.0,"Johnston, Mr. Andrew G",male,NaN,1.0,2.0,W./C. 6607,23.45,NaN,S
104,105,No,3.0,"Gustafsson, Mr. Anders Vilhelm",male,37.0,2.0,0.0,3101276,7.925,NaN,S
144,145,No,2.0,"Andrew, Mr. Edgardo Samuel",male,18.0,0.0,0.0,231945,11.5,NaN,S
327,328,Yes,2.0,"Ball, Mrs. (Ada E Hall)",female,36.0,0.0,0.0,28551,13,D,S
63,64,No,3.0,"Skoog, Master. Harald",male,4.0,3.0,2.0,347088,27.9,NaN,S


The .sample() method randomly selects rows from the dataset, providing a more representative view of the data than just the first few rows.

### Question 1: Based on the output above, what is the main difference between .head and .sample? And why would you use .sample over .head?

In general, they both do similar things. Both display a certain number of records, and both accept a number to specify how many records to show. However, the main difference is that .head() always shows the **first X records**, while .sample() shows **X random records**. This gives you a more valid sample and accurate representation of the entire dataset, not just the beginning.

We've also seen another method to explore our dataset: looking at the data types.

In [34]:
titanic.dtypes

PassengerId      int64
Survived        object
Pclass         float64
Name            object
Sex             object
Age            float64
SibSp          float64
Parch          float64
Ticket          object
Fare            object
Cabin           object
Embarked        object
dtype: object

Checking data types with .dtypes helps us see how pandas interprets each column. Numeric types like int64 or float64 are ideal for calculations, while 'object' types are usually text strings that might need conversion to numbers.

### Question 2: Looking at the datatypes of each column, which ones will we likely need to adapt?

We always want as many numeric values as possible for analysis. Integers and floats are ideal data types, while 'object' types (usually strings) are less useful and may need to be converted.

The columns that likely need adaptation are:

- Survived

- Name

- Sex

- Age

- Ticket

- Cabin

- Embarked

Another method to explore our dataset is by using the .describe() function. This provides statistical summaries, such as the minimum value, maximum value, average value, and more!

In [35]:
titanic.describe()

,PassengerId,Pclass,Age,SibSp,Parch
count,892.000000,891.000000,714.000000,891.000000,891.000000
mean,446.500000,2.308642,29.706120,0.523008,0.381594
std,257.642517,0.836071,14.523986,1.102743,0.806057
min,1.000000,1.000000,0.420000,0.000000,0.000000
25%,223.750000,2.000000,20.125000,0.000000,0.000000
50%,446.500000,3.000000,28.000000,0.000000,0.000000
75%,669.250000,3.000000,38.000000,1.000000,0.000000
max,892.000000,3.000000,80.000000,8.000000,6.000000


The .describe() method gives summary statistics for numeric columns, including count, mean, standard deviation, minimum, maximum, and quartiles. This helps us understand the distribution of our data.

Similar to how we create Series of booleans for selecting data subsets, we can create entire DataFrames of booleans! The most useful case is identifying missing values.

In [36]:
titanic.isna()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,False,False,False,False,False,True,False,False,False,False,True,False
1,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,True,False
3,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...
887,False,False,False,False,False,False,False,False,False,False,False,False
888,False,False,False,False,False,True,False,False,False,False,True,False
889,False,False,False,False,False,False,False,False,False,False,False,False
890,False,False,False,False,False,False,False,False,False,False,True,False


The .isna() method returns a DataFrame where each cell is True if the original value is missing (NaN) and False otherwise.

But what do we do with this information? Well, it's simple! When we combine it with our earlier .describe() method, we get slightly different behavior that gives us a great overview of the missing values!

In [37]:
titanic.isna().describe()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
count,892,892,892,892,892,892,892,892,892,892,892,892
unique,1,2,2,2,1,2,2,2,2,2,2,2
top,False,False,False,False,False,False,False,False,False,False,True,False
freq,892,891,891,891,892,714,891,891,891,891,688,890


By applying .describe() to the boolean DataFrame from .isna(), we get counts of True (missing) and False (present) values per column, helping us quantify missing data.

- **count:** The total number of rows in the dataset (e.g., 892 for the Titanic dataset). This is the same for every column since .describe() is applied to the full DataFrame.

- **unique:** The number of unique values in the column. For boolean data (True/False), this is usually 2 (True and False), but it could be 1 if a column has no missing values (only False) or is entirely missing (only True).

- **top:** The most common value in the column. For missing data analysis, this is often False (meaning most values are present) or True (meaning most values are missing).

- **freq:** The frequency (count) of the top value. This tells you how many rows have the most common state (e.g., how many are missing or present).

### Question 3: Try to interpret the output above. Where do the most missing values lie? Are there columns with no missing values?

The most missing values are in the 'Cabin' column. You can see this because it's the only column where 'True' (indicating a missing value from .isna()) is the most common value. Looking at the frequency, there are 687 missing values in this column.

The columns with no missing values are PassengerId and Sex, as they both have only 'False' values.

## 3.3 First data transformations 

Now that we've discussed modifying data types, let's start doing it!

- The easiest and most common way is called **one-hot encoding**, also known as **dummy variables**. 

- The idea is simple: for a binary variable (either A or B), we can represent it with the numbers 0 and 1.

- This column represents **whether the variable is present**. It's important to decide what "present" means for our interpretation.

- In the Titanic dataset, a good candidate is the 'Survived' column. We can define survival as the variable being present, so 'Yes' becomes 1 and 'No' becomes 0.

How do we implement this? Manually replacing each value would take way too long. Fortunately, pandas has the .map() function built-in.

We call the column we want to map and pass a dictionary where the keys are the old values and the values are the new ones.

In [38]:
titanic['Survived'] = titanic['Survived'].map({'No': 0, 'Yes' : 1})

The .map() method applies a dictionary mapping to replace values in the Series. Here, it converts 'No' to 0 and 'Yes' to 1.

In [39]:
titanic['Survived']

0      0.0
1      1.0
2      1.0
3      1.0
4      0.0
      ... 
887    1.0
888    0.0
889    1.0
890    0.0
891    0.0
Name: Survived, Length: 892, dtype: float64

### Question 4: One-hot encoding works by indicating whether a certain variable is present, thus limiting it to binary columns only. This means we cannot simply apply it to columns like 'Embarked'. Can you think of a way to use this dummy variables technique anyways to encode such columns? Are there any new limitations for this?

A column like 'Embarked' has multiple possible values. While we cannot convert it to a single column indicating presence, we can create **a new column for each possible value** and mark if that value is present for each record. We must ensure this transformation makes sense.

For example, in a car dataset, applying this to the model name column would not add value, since each car has a unique model. Each new column would have mostly zeros and one 1, increasing the dataset size without benefit. We'll discuss this more in the Introduction to Machine Learning class. However, it could be useful for the brand name column, where many cars share the same brand, providing real information gain.

Pandas provides the .get_dummies() function for this.

In [40]:
titanic = pd.get_dummies(titanic, columns = ['Embarked'])

pd.get_dummies() creates new binary columns for each unique value in the specified column(s). Here, it adds columns like 'Embarked_C', 'Embarked_Q', 'Embarked_S' with 1 where the value matches and 0 otherwise.

In [41]:
titanic.sample(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked_C,Embarked_Q,Embarked_S
697,698,1.0,3.0,"Mullens, Miss. Katherine \Katie\""""",female,NaN,0.0,0.0,35852,77.333,NaN,False,True,False
474,475,0.0,3.0,"Strandberg, Miss. Ida Sofia",female,22.0,0.0,0.0,7553,98.375,NaN,False,False,True
703,704,0.0,3.0,"Gallagher, Mr. Martin",male,25.0,0.0,0.0,36864,77.417,NaN,False,True,False
754,755,1.0,2.0,"Herman, Mrs. Samuel (Jane Laver)",female,48.0,1.0,2.0,220845,65,NaN,False,False,True
350,351,0.0,3.0,"Odahl, Mr. Nils Martin",male,23.0,0.0,0.0,7267,9.225,NaN,False,False,True


Great! But are these the only ways to transform data? Of course not! Sometimes datasets already use booleans for binary data. Pandas makes it easy to convert data types, as long as they are compatible (like True/False and 1/0).

In [42]:
titanic['Survived'].astype('bool')

0      False
1       True
2       True
3       True
4      False
       ...  
887     True
888    False
889     True
890    False
891    False
Name: Survived, Length: 892, dtype: bool

The **.astype('bool')** method converts the Series to boolean type:
-  changing 0 to -> False
-  1 to -> True.

If you ever lose track of the current data type, just check it!

In [43]:
titanic['Survived'].dtype

dtype('float64')

Accessing .dtype on a Series shows its data type, helping us confirm changes.

## 3.4 Missing values

There are three main ways to handle missing data:

- Removing data

- Data imputation

- Data flagging

Removing data is the simplest, but it leads to the most information loss.

In [44]:
titanic2 = titanic.dropna()

dropna() removes any row containing at least one missing value, which can drastically reduce the dataset size.

In [45]:
titanic.isna().describe()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked_C,Embarked_Q,Embarked_S
count,892,892,892,892,892,892,892,892,892,892,892,892,892,892
unique,1,2,2,2,1,2,2,2,2,2,2,1,1,1
top,False,False,False,False,False,False,False,False,False,False,True,False,False,False
freq,892,891,891,891,892,714,891,891,891,891,688,892,892,892


In [46]:
titanic2.isna().describe()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked_C,Embarked_Q,Embarked_S
count,185,185,185,185,185,185,185,185,185,185,185,185,185,185
unique,1,1,1,1,1,1,1,1,1,1,1,1,1,1
top,False,False,False,False,False,False,False,False,False,False,False,False,False,False
freq,185,185,185,185,185,185,185,185,185,185,185,185,185,185


Notice how easy it was to drop all missing data, but it left us with only 185 rows out of 891 – a huge loss! By default, .dropna() deletes any row with a missing value. To adjust this, we can set a threshold.

In [47]:
titanic2 = titanic.dropna(thresh=5)

dropna(thresh=5) keeps only rows with at least 5 non-missing values, providing more control over data removal.

In [48]:
titanic2.isna().describe()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked_C,Embarked_Q,Embarked_S
count,892,892,892,892,892,892,892,892,892,892,892,892,892,892
unique,1,2,2,2,1,2,2,2,2,2,2,1,1,1
top,False,False,False,False,False,False,False,False,False,False,True,False,False,False
freq,892,891,891,891,892,714,891,891,891,891,688,892,892,892


Now we haven't removed any rows with missing values. You can see how the threshold can be tricky to tune!

Another way to handle missing values is by filling them with reasonable estimates. There are several built-in functions for this!

In [49]:
titanic['Age'].value_counts()

Age
24.00    30
18.00    26
22.00    26
28.00    25
19.00    25
         ..
24.50     1
0.67      1
0.42      1
34.50     1
74.00     1
Name: count, Length: 88, dtype: int64

In [50]:
age = titanic['Age'].ffill()
age.value_counts()

Age
24.00    39
19.00    38
21.00    33
28.00    32
22.00    31
         ..
24.50     1
0.67      1
0.42      1
34.50     1
74.00     1
Name: count, Length: 88, dtype: int64

ffill() fills missing values with the last known (previous) value in the column.

In [51]:
age = titanic['Age'].bfill()
age.value_counts()

Age
24.00    38
18.00    35
22.00    30
28.00    30
25.00    30
         ..
24.50     1
0.67      1
0.42      1
34.50     1
74.00     1
Name: count, Length: 88, dtype: int64

bfill() fills missing values with the next known (subsequent) value in the column.

Sometimes built-in functions aren't suitable, so you need a custom approach.

In [52]:
age = titanic['Age'].fillna(titanic['Age'].median())
age.value_counts()

Age
28.00    203
24.00     30
22.00     26
18.00     26
19.00     25
        ... 
24.50      1
0.67       1
0.42       1
34.50      1
74.00      1
Name: count, Length: 88, dtype: int64

fillna() with the median fills missing values with the middle value of the non-missing data, preserving the central tendency.

Remember to be very careful when filling data like this! It's easy for a value to seem good but actually distort the dataset. That's why sometimes it's better to 'flag' missing values by assigning a special value that doesn't interfere with the data.

Make sure to choose values that won't ruin your dataset – avoid extremes like very low or high numbers!

In [53]:
age = titanic['Age'].fillna(0)
age.value_counts()

Age
0.00     178
24.00     30
18.00     26
22.00     26
28.00     25
        ... 
24.50      1
0.67       1
0.42       1
34.50      1
74.00      1
Name: count, Length: 89, dtype: int64

Filling with 0 flags missing values but might skew statistics if 0 is a valid age.

In [54]:
titanic['Age_missing'] = titanic['Age'].isna().astype(int)
age = titanic['Age'].fillna(titanic['Age'].median())  # Or another imputation method
age

0      28.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
887    19.0
888    28.0
889    26.0
890    32.0
891    27.0
Name: Age, Length: 892, dtype: float64

In [55]:
titanic

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked_C,Embarked_Q,Embarked_S,Age_missing
0,1,0.0,3.0,"Braund, Mr. Owen Harris",male,NaN,1.0,0.0,A/5 21171,7.25,NaN,False,False,True,1
1,2,1.0,1.0,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1.0,0.0,PC 17599,712.833,C85,True,False,False,0
2,3,1.0,3.0,"Heikkinen, Miss. Laina",female,26.0,0.0,0.0,STON/O2. 3101282,7.925,NaN,False,False,True,0
3,4,1.0,1.0,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1.0,0.0,113803,53.1,C123,False,False,True,0
4,5,0.0,3.0,"Allen, Mr. William Henry",male,35.0,0.0,0.0,373450,8.05,NaN,False,False,True,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
887,888,1.0,1.0,"Graham, Miss. Margaret Edith",female,19.0,0.0,0.0,112053,30,B42,False,False,True,0
888,889,0.0,3.0,"Johnston, Miss. Catherine Helen \Carrie\""""",female,NaN,1.0,2.0,W./C. 6607,23.45,NaN,False,False,True,1
889,890,1.0,1.0,"Behr, Mr. Karl Howell",male,26.0,0.0,0.0,111369,30,C148,True,False,False,0
890,891,0.0,3.0,"Dooley, Mr. Patrick",male,32.0,0.0,0.0,370376,7.75,NaN,False,True,False,0
